source/fundamentals/connection/tuning.txt

```diff
@@ -0,0 +1,258 @@
+after it completes a successful monitoring check. The driver creates these connections ``maxConnecting``
+connections at a time. This happens in the background and is not blocking.
+
+.. example::
+
+   With a ``minPoolSize`` of **10**, the driver establishes 10 connections.
+   If each connection takes ~50 ms to establish, the driver establishes
+   all connections after ~250 ms. That is, 50 ms * 10 (the ``minPoolSize``)
+   / 2 (``maxConnecting``).
+
+
+Growth
+++++++
+
+.. note:: Server Monitoring
+
+   The driver establishes monitoring connections to continuously check
+   deployment health. These connections are not managed by the connection
+   pool.
+
+The driver establishes connections until ``minPoolSize`` connections are
+available in the connection pool. As the driver processes operations,
+connections are checked in and out. Operations that aren't immediately
+assigned a connection from the connection pool enter a wait queue, and
+are processed when a connection becomes available.
+
+If there are operations in the wait queue **and** the number of connections
+in the pool is less than the ``maxPoolSize``, the driver establishes connections
+to process queued operations, up to the ``maxPoolSize``.
+
+.. example::
+
+   Your ``minPoolSize`` is **4** and your ``maxPoolSize`` is **10**. In
+   response to a successful email campaign, thousands of users log in to
+   your application. The driver uses the four connections in the pool,
+   and places other operations in the wait queue.
+
+   In response to the wait queue growth, the driver establishes more
+   connections to process operations, ``maxConnecting`` at a time until the
+   connection pool reaches ``maxPoolSize`` connections.
+
+Reduction
++++++++++
+
+Use ``maxIdleTimeMS`` to reduce the number of connections in the pool to
+reduce resource requirements both on the application and on the database
+machines. The driver has no default, but 30 minutes is a good starting point.
+
+Connection Storms
+-----------------
+
+A **connection storm** is an event in which application servers
+start an increasing number of connection requests, but the host server
+CPU is unable to schedule them.
+
+.. example::
+
+   Consider an application that uses MongoDB, and is served from an
+   auto scaling service. A sudden spike in traffic could cause the scaling
+   service to stand up many new application instances. This sudden burst
+   in application instances attempting to establish connections to MongoDB
+   is one example of a connection storm.
+
+Common Causes
++++++++++++++
+
+Common causes for connection storms in MongoDB include:
+
+- Incorrect application behavior
+- Network latency
+- Insufficient hardware for the workload
+- Failover events
+
+In MongoDB, connection storms can have multiple effects. For example:
+
+- The high volume of connection requests leads to latency in processing.
+- Replication connections between nodes may be blocked or slowed, so replication
```

Comment:
I think this has been fixed in recent versions of MongoDB, but you'll have to verify with the server team / tses.

Reply:
I'm asking someone from Server who recently reviewed some performance work to look at this.

Reply:
I suggest an alternate framing here, which peels out the activity (Rate of connections vs. Total number of connections) and then describes the possible effects of either of them together (which are similar), with less technical detail about how the effects can unfold. Should help future-proof these docs as improvements are made over time:

---

The effects of connection storms follow from either:

* the performance impact of the rate at which new connections are being made. Rapidly incoming connection requests can cause bottlenecks during connection establishment. Since it is common to retry failed connections, connection storms can persist until some load is shed.
* inability to connect to a database node when that node's [maxIncomingConnections|https://docs.mongodb.com/manual/reference/configuration-options/#mongodb-setting-net.maxIncomingConnections] setting is reached.

Either of these concerns can lead to practical issues like:

* Hardware overuse, including heavy queuing, crashing and restarting due to lack of memory, etc.
* Replica set failover due to interference with intra-node system operations.
* Inability to connect to and diagnose an issue on the database side.
* Loss of connectivity for monitoring tools and other support services.
* Interference with monitoring and administration activities.